aten/src/ATen/native/ConvolutionMM3d.cpp

```diff
@@ -258,28 +259,32 @@ void slow_conv3d_backward_update_grad_input_frame(
     int64_t pad_height,
     int64_t pad_width,
     int64_t groups) {
-  if (groups > 1) {
-    auto n = grad_output.size(0);
-    auto d = grad_output.size(1);
-    auto h = grad_output.size(2);
-    auto w = grad_output.size(3);
-    auto grad_output_2d = grad_output.reshape({groups, n / groups, d * h * w});
-    auto weight_g =
-        weight.reshape({groups, weight.size(0), weight.size(1) / groups});
-    auto fgrad_input_g = fgrad_input.reshape(
-        {groups, fgrad_input.size(0) / groups, fgrad_input.size(1)});
-
-    at::bmm_out(fgrad_input_g, weight_g, grad_output_2d);
-  } else {
-    auto grad_output_2d = grad_output.reshape(
-        {grad_output.size(0),
-         grad_output.size(1) * grad_output.size(2) * grad_output.size(3)});
-    at::mm_out(fgrad_input, weight, grad_output_2d);
+  // Compute fgrad_input = weight.T * grad_output.reshape({grad_output.shape(0), -1})
+  // Note gemm expects fortran order, so all 3 matrices are transposed.
+  // Swapping argument order cancels this, since C == AB <=> T(C) == T(B)T(A)
+  const int64_t m = grad_output.size(1) * grad_output.size(2) * grad_output.size(3);
+  const int64_t n = weight.size(1);
+  const int64_t k = weight.size(0) / groups;
+
+  const int64_t lda = m;
+  const int64_t ldb = n;
+  const int64_t ldc = m;
+
+  for (int64_t group = 0; group < groups; ++group) {
```

Comment:
mkl batched gemm (that was used for the grouped case) was an important perf optimization that someone needed. Can we bring it back here? It's not very convenient as it's not part of `cpublas`, but I don't feel comfortable just removing it.

Reply:
Sure, It might make sense to add a batched gemm to `cpublas`. I'll see how it goes.